File: uberPaste/src/main/scala/uber/paste/startup/StartupListener.scala
/**
 * Copyright (C) 2010 alex <me@alex.0x08.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package uber.paste.startup


import javax.servlet.{ServletContextListener, ServletContext, ServletContextEvent}
import org.springframework.context.ApplicationContext
import org.apache.commons.io.FileUtils
import org.springframework.web.context.support.WebApplicationContextUtils
import uber.paste.model._
import uber.paste.dao._
import uber.paste.build._
import org.compass.gps.CompassGps;
import java.io.IOException
import uber.paste.dao.UserExistsException
import uber.paste.base.{MergedPropertyConfigurer, Loggered, SystemInfo}
import uber.paste.build._
import uber.paste.mail.EmbeddedSMTPServer
import java.nio.file._
import java.nio.file.attribute.BasicFileAttributes
import uber.paste.manager.PasteManager
import scala.collection.JavaConversions._

/*
class AdvEventLoadListener extends DefaultPostLoadEventListener with Loggered{


  override def onPostLoad(p1: PostLoadEvent) {
    logger.debug("_on Load call for class "+p1.getEntity.getClass.getName)


    super.onPostLoad(p1)

    if (p1.getEntity.isInstanceOf[Paste]==false) {
        return
    }

      val obj:Paste =p1.getEntity.asInstanceOf[Paste]

      obj.setTagsAsString({
        val out =new StringBuilder
        for (s<-obj.getTags()) {
          out.append(s).append(" ")
        }
        out.toString })

      logger.debug("filled tagsAsString "+obj.getTagsAsString())
    }

}
*/
class StartupListener extends ServletContextListener with Loggered{
  
  override def contextInitialized( event:ServletContextEvent): Unit = {

    val context:ServletContext = event.getServletContext()

    val  ctx:ApplicationContext =
      WebApplicationContextUtils.getRequiredWebApplicationContext(context)

  /*  val sf:SessionFactory = ctx.getBean("sessionFactory").asInstanceOf[SessionFactory]

    val listeners:Array[PostLoadEventListener] = new Array(1)
    //listeners(0) =new DefaultPostLoadEventListener
    listeners(0) =new AdvEventLoadListener()


    sf.asInstanceOf[SessionFactoryImpl].getEventListeners.setPostLoadEventListeners(listeners)
    */

    System.setProperty ("jsse.enableSNIExtension", "false")
    
    var props:MergedPropertyConfigurer =  ctx.getBean("propertyConfigurer").asInstanceOf[MergedPropertyConfigurer]


    val configDao:ConfigDao = ctx.getBean("configDao").asInstanceOf[ConfigDao]
    val pasteDao:PasteDao = ctx.getBean("pasteDao").asInstanceOf[PasteDao]
    val projectDao:ProjectDao = ctx.getBean("projectDao").asInstanceOf[ProjectDao]


    if (configDao.isPropertySet(ConfigProperty.IS_INSTALLED.getCode, "1")) {
      
      val dbVersion = new AppVersion().fillFromConfigProperty(
        configDao.getProperty(ConfigProperty.APP_VERSION.getCode))
      
           
            if (dbVersion != null) {
      
        logger.debug("__currentSettings:" + dbVersion.getFull)
                
                val check =  SystemInfo.instance.getRuntimeVersion().compareTo(dbVersion)
                
                check match {
           
                    case 0 => {
                        logger.info("Application and db versions match.")
                      }
                    case 1 => {
                        logger.warn("DB version is older than application: "
                                +dbVersion.getFull()+" | "+SystemInfo.instance.getRuntimeVersion().getFull()+". You can get problems.")
                    }
                    case -1 =>   {
                        logger.warn("Application version is older than database: "
                                +dbVersion.getFull()+" | "+SystemInfo.instance.getRuntimeVersion().getFull()+". You can get problems.")
                    }
                     
                    case _ => {
                    logger.error("Uncomparable db and application versios: "
                             +dbVersion.getFull()+" | "+SystemInfo.instance.getRuntimeVersion().getFull()+". Cannot continue.")
                    }    
                        
                }
                                
                
                 logger.info("Loading application ver. "+SystemInfo.instance.getRuntimeVersion().getFull()
                             +" db ver. "+dbVersion.getFull())
               
            }
      
      
      logger.info("Database already created. skipping db generation stage..")

        SystemInfo.instance.setProject(projectDao.getLast)
        
      reindex(ctx,props)


      if (props.getProperty("config.load.data.enabled").equals("1")) {

        logger.info("Data loading enabled..")

        var dirs = props.getProperty("config.load.data.folders").split(",")
        
        
         for (f<-dirs) {
            logger.info("Start loading for "+f)

             new Thread(new Runnable() {
        def run() {
          Files.walkFileTree(Paths.get(f), new WalkFileTree(pasteDao))
        }
      }).start()

        }
      }
 
      
      
      
      startSmtpServer(ctx,props)

      for(p<-Priority.list) {
        
        PasteManager.Stats.valueOf(p.getCode).increment(pasteDao.countAll(p).toInt)
        
      }
      
     return
    }


    val userDao:UserDao = ctx.getBean("userDao").asInstanceOf[UserDao]


    try {

      
      val admin = userDao.save(
        UserBuilder.createNew()
        .addRole(Role.ROLE_ADMIN)
        .addUsername("admin")
        .addPassword("admin")
        .addName("Default")
        .get())


      val alex = userDao.save(
        UserBuilder.createNew()
        .addRole(Role.ROLE_USER)
        .addUsername("alex")
        .addPassword("login")
        .addName("Alex")
        .get());


      var p1:Paste = new Paste
      p1.setOwner(admin)
      p1.setName("first test record")
      p1.setText("test record body")
     
     
      var p2:Paste = new Paste
      p2.setOwner(alex)
      p2.setName("second test record")
      p2.setText("test second record body")
     
      pasteDao.save(p1)
      pasteDao.save(p2)

      val project = new Project
      project.setName("Sample project")
      project.setDescription("Full project description")
      
      projectDao.persist(project)
        
      configDao.persist(ConfigProperty.IS_INSTALLED)
      configDao.persist(ConfigProperty.UPLOADS_DIR)
      
      
      ConfigProperty.APP_VERSION.setValue(SystemInfo.instance.getRuntimeVersion.toDbString)
      configDao.persist(ConfigProperty.APP_VERSION)

      logger.debug("saved version "+ConfigProperty.APP_VERSION.getValue)

      reindex(ctx,props)

      logger.info("db generation completed successfully.")

      startSmtpServer(ctx,props)

    } catch {
      case e:UserExistsException => {
          logger.error(e.getLocalizedMessage,e)

        } case e:IOException => {
          logger.error(e.getLocalizedMessage,e)

        }
    }  
   
  }

  override def contextDestroyed(servletContextEvent:ServletContextEvent):Unit = {
    
  }


  def startSmtpServer(ctx:ApplicationContext,props:MergedPropertyConfigurer) {

    if (props.getProperty("config.smtpd.enabled").equals("1")) {
      EmbeddedSMTPServer.createInstance(ctx).start()
    } else {
      logger.info("smtpd was disabled. skipping it.")
    }
  }

  def reindex( ctx:ApplicationContext,props:MergedPropertyConfigurer):Unit = {
    if (props.getProperty("config.reindex.enabled").equals("1")) {
      val compassGps:CompassGps = ctx.getBean("compassGps").asInstanceOf[CompassGps]
      compassGps.index()
      logger.info("reindex completed.")
    } else {
      logger.info("reindex was disabled. skipping it.")
    }

  }


  class WalkFileTree(pasteDao:PasteDao) extends SimpleFileVisitor[Path] with Loggered {

    override def preVisitDirectory(dir:Path, attr:BasicFileAttributes):FileVisitResult= {
    //  System.out.printf("Begin Directory: %s%n", dir)
      return FileVisitResult.CONTINUE
    }

    // Print information about each file/symlink visited.
    override def visitFile(file:Path, attr:BasicFileAttributes):FileVisitResult = {

      //logger.debug("visiting file: "+file.getFileName()+" is java "+file.getFileName().endsWith(".java"))


      if (file.getFileName.toString.endsWith(".java")) {

          logger.debug("pasting file: "+file)

         val p:Paste = new Paste
         p.setCodeType(CodeType.Java)
         p.setName(file.getFileName.toString)
         p.setText(FileUtils.readFileToString(file.toFile))

        try {

          pasteDao.save(p)

        }  catch {
          case e:Exception => {
            logger.error(e.getLocalizedMessage,e)
          }
        }

       }
      return FileVisitResult.CONTINUE
    }

    // If there is an error accessing the file, print a message and continue.
    override def visitFileFailed(file:Path, ex:IOException):FileVisitResult= {
      logger.error(ex.getLocalizedMessage,ex)
      return FileVisitResult.CONTINUE;  // or TERMINATE
    }


  }

}

File: shareService/webapp/src/main/java/uber/megashare/listener/StartupListener.java
/**
 * Copyright (C) 2011 aachernyshev <alex@0x08.tk>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uber.megashare.listener;

import java.io.File;
import java.util.Calendar;
import java.util.List;
import javax.servlet.ServletContext;
import javax.servlet.ServletContextEvent;
import javax.servlet.ServletContextListener;
import org.apache.lucene.queryParser.ParseException;
import org.springframework.context.ApplicationContext;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.context.support.WebApplicationContextUtils;
import uber.megashare.base.LoggedClass;
import uber.megashare.base.SettingsBuilder;
import uber.megashare.dao.SharedFileDao;
import uber.megashare.dao.UserDao;
import uber.megashare.model.Role;
import uber.megashare.model.User;
import uber.megashare.service.SettingsManager;
import uber.megashare.service.UserManager;
import org.neo4j.graphdb.traversal.Evaluators;
import org.neo4j.graphdb.traversal.TraversalDescription;
import org.neo4j.kernel.Traversal;
import org.neo4j.kernel.Uniqueness;
import uber.megashare.base.SystemInfo;
import uber.megashare.dao.ProjectDao;
import uber.megashare.model.Project;
import uber.megashare.model.SystemProperties;
import uber.megashare.model.tree.FolderNode;
import uber.megashare.model.tree.NodeType;
import uber.megashare.model.tree.Rels;
import uber.megashare.service.tree.FolderService;
/**
 * <p>StartupListener class used to initialize and database settings and
 * populate any application-wide drop-downs.
 * <p/>
 * <p>Keep in mind that this listener is executed outside of
 * OpenSessionInViewFilter, so if you're using Hibernate you'll have to
 * explicitly initialize all loaded data at the GenericDao or service level to
 * avoid LazyInitializationException. Hibernate.initialize() works well for
 * doing this.
 *
 * @author <a href="mailto:matt@raibledesigns.com">Matt Raible</a>
 */
public class StartupListener extends LoggedClass implements ServletContextListener {

   
    
    /**
     *
     */
    private static final long serialVersionUID = -132229174806241432L;
    
    private UserDao userDao;
    
    private SharedFileDao shareDao;
    
    private ProjectDao projectDao;

    /**
     * {@inheritDoc}
     */
    @Override
    public void contextInitialized(ServletContextEvent event) {
        getLogger().debug("Initializing context...");

        ServletContext context = event.getServletContext();
        setupContext(context);
    }

    /**
     * This method uses the LookupManager to lookup available roles from the
     * data layer.
     *
     * @param context The servlet context
     */
    private void setupContext(ServletContext context) {
        try {
            ApplicationContext ctx = WebApplicationContextUtils.getRequiredWebApplicationContext(context);

          

            User startup = new User();
            startup.setName("DB Startup");
            startup.setLogin("startup");
            startup.setPassword("startup");
            startup.getRoles().add(Role.ROLE_ADMIN);

            // log user in automatically
            UsernamePasswordAuthenticationToken auth = new UsernamePasswordAuthenticationToken(
                    "startup", "startup", startup.getAuthorities());
            auth.setDetails(startup);
            SecurityContextHolder.getContext().setAuthentication(auth);


            SettingsManager settingsDao = (SettingsManager) ctx.getBean("settingsManager");
            UserManager userManager = (UserManager) ctx.getBean("userManager");
            userDao = (UserDao) ctx.getBean("userDao");
            shareDao = (SharedFileDao) ctx.getBean("shareDao");
            projectDao = (ProjectDao) ctx.getBean("projectDao");
            

    //        UserCustomExtendedRepository repo = (UserCustomExtendedRepository) ctx.getBean("userCustomExtendedRepository");

            // EmbeddedSMTPServer.createInstance(ctx).start();


            SystemProperties sp = settingsDao.getCurrentSettings();
            getLogger().debug("__currentSettings:" + sp);
           
            if (sp != null && sp.getInstallDate() != null) {
                
                int check =  SystemInfo.getInstance().getRuntimeVersion().compareTo(sp.getAppVersion());
                
                switch(check) {
           
                    case 0:
                        getLogger().info("Application and db versions match.");
                        break;
                    case 1:
                        getLogger().warn("DB version is older than application: "
                                +sp.getAppVersion().getImplVersionFull()+" | "+SystemInfo.getInstance().getRuntimeVersion().getImplVersionFull()+". You can get problems.");
                        break;
                        
                    case -1:    
                        getLogger().warn("Application version is older than database: "
                                +sp.getAppVersion().getImplVersionFull()+" | "+SystemInfo.getInstance().getRuntimeVersion().getImplVersionFull()+". You can get problems.");
                        break;
                    
                    case -2:
                    default:    
                    getLogger().error("Uncomparable db and application versios: "
                             +sp.getAppVersion().getImplVersionFull()+" | "+SystemInfo.getInstance().getRuntimeVersion().getImplVersionFull()+". Cannot continue.");
           
                        
                }
                                
                
                 getLogger().info("Loading application ver. "+SystemInfo.getInstance().getRuntimeVersion().getImplVersionFull()+" db ver. "+sp.getAppVersion().getImplVersionFull());
               
                
                reindex();

                getLogger().info("Database already set. Aborting creation..");

                return;
            }

            sp =settingsDao.merge(SettingsBuilder.getInstance() //install date
                    .addInstallDate(Calendar.getInstance().getTime())
                    .addAppVersion()                    
                    .addUploadDir("files")
                    .get());

            SystemInfo.getInstance().setRuntimeVersion(sp.getAppDbVersion());
                
            
            getLogger().info("Loading application ver. "+sp.getAppVersion().getImplVersionFull());
            
            File uf = new File(settingsDao.getCurrentSettings().getUploadDir());
            if (!uf.exists()) {

                if (!uf.mkdirs()) {
                    throw new IllegalStateException("Cannot create directory " + uf.getAbsolutePath());
                }
            }

            Project test_p = new Project();
            test_p.setDescription("Sample project description");
            test_p.setName("Sample Project");
            
            test_p = projectDao.saveObject(test_p);
            
             Project test_p2 = new Project();
            test_p2.setDescription("Sample project2 description");
            test_p2.setName("Sample Project2");
           
            test_p2 = projectDao.saveObject(test_p2);
           
            
            User admin = new User();
            admin.setName("Administrator");
            admin.setLogin("admin");
            admin.setRelatedProject(test_p);
            admin.getRoles().add(Role.ROLE_ADMIN);

            admin = userManager.changePassword(admin, "admin");
            admin = userDao.saveObject(admin);
        
        


            User alex = new User();
            alex.setName("Alex");
            alex.setLogin("alex");
            alex.setEmail("alex@0x08.tk");
            alex.getRoles().add(Role.ROLE_USER);
            alex.setRelatedProject(test_p2);
            alex = userManager.changePassword(alex, "login");
            alex = userDao.saveObject(alex);

            getLogger().info("alex=" + alex.getLogin() + "|" + alex.getPassword());


            List<User> found_users;
         
                found_users = userDao.search("name:" + alex.getName());

                getLogger().info("Found " + found_users.size() + " users containing login : " + alex.getLogin());
                for (User u : found_users) {
                    getLogger().info("login=" + u.getLogin() + "|" + u.getPassword());

                }
                
                
               FolderService folderManager = (FolderService)ctx.getBean("folderService");
               
               FolderNode defaultParent = new FolderNode();
                          defaultParent.setName("Default parent");
                          defaultParent.setType(NodeType.PARENT);
                          
                        defaultParent =  folderManager.save(defaultParent);
                
                        System.out.println("_parent id: "+defaultParent.getId()+" type "+defaultParent.getType());
                        
                         FolderNode sub1 = new FolderNode();
                          sub1.setName("Sub 1");
                          sub1.setType(NodeType.FOLDER);
                          
                          
                        FolderNode sub2 = new FolderNode();
                          sub2.setName("Sub 2");
                         sub2.setType(NodeType.FOLDER);
                          
                          sub1.setParent(defaultParent);
                          
                        /*  Set<FolderNode> ch = new HashSet<>();
                          ch.add(sub2);
                          sub1.setChildren(ch);
                          */
                          
                          
                          
                          sub1 =  folderManager.save(sub1);

                          sub2.setParent(sub1);

                          sub2 =  folderManager.save(sub2);

                          
                        System.out.println("_sub1 id: "+sub1.getId()+" type "+sub1.getType()+" sub2 id: "+sub2.getId()+" sub2 parent "+sub2.getParent().getId());
                    
            
          final TraversalDescription FRIENDS_TRAVERSAL = Traversal.description()
        .depthFirst()
        .relationships( Rels.CHILD )
         .evaluator(Evaluators.includingDepths(1, 2))
        .uniqueness( Uniqueness.RELATIONSHIP_GLOBAL );
               
          for (FolderNode friend : folderManager.findAllByQuery("START a=node(1)\n" +
"MATCH a-[:CHILD*1..3]->x\n" +
"RETURN a,x", uf)) {
    
              System.out.println("node from parent: " + friend.getId());

          }
          
          
          /* for (FolderNode friend : folderManager.findAllByTraversal(sub1,FRIENDS_TRAVERSAL)) {
    
              System.out.println("node from sub1: " + friend.getId());

          }*/
          
        
          /*
            for (FolderNode f : folderManager.findAllByPropertyValue("type", NodeType.PARENT)) {
                    System.out.println("parent node: " + f.getName());
            
                for (FolderNode ff : f.getChildren()) {
                    System.out.println("sub nodes : " + ff.getName());
                }
            }
            */ 
             

   
                getLogger().info("__done creation");
   
            } catch (ParseException ex) {
                getLogger().error(ex.getLocalizedMessage(), ex);
         
        }

    }

    private void reindex() {

        getLogger().info("Reindexing..");

        userDao.indexAll();
        shareDao.indexAll();

    }

    /**
     * Shutdown servlet context (currently a no-op method).
     *
     * @param servletContextEvent The servlet context event
     */
    //  @Override
    @Override
    public void contextDestroyed(ServletContextEvent servletContextEvent) {
     
    }

   
}
